src/MergeContext.js

```diff
@@ -198,19 +217,233 @@ class StatusChecks
+class PrState
+{
+    // treat as private; use static methods below instead
+    constructor(state) {
+        this._state = state;
+    }
+
+    static Brewing() { return new PrState(-1); }
+    static Staged() { return new PrState(0); }
+    static Merged() { return new PrState(1); }
+
+    brewing() { return this._state < 0; }
+    staged() { return this._state === 0; }
+    merged() { return this._state > 0; }
+
+    toString() {
+        if (this.brewing())
+            return "brewing";
+        if (this.staged())
+            return "staged";
+        assert(this.merged());
+        return "merged";
+    }
+}
+
+// Determines the feature branch position relative to its base branch: ahead,
+// merged, or diverged. This class essentially decides whether/which one of
+// the two branches is fully contained in another.
+class BranchPosition
+{
+    constructor(baseRef, featureRef) {
+        assert(baseRef !== featureRef);
+        this._baseRef = baseRef;
+        this._featureRef = featureRef;
+        this._status = null;
+    }
+
+    async compute() {
+        this._status = await GH.compareCommits(this._baseRef, this._featureRef);
+    }
+
+    // feature > base:
+    // the feature branch contains the base branch and some additional commits
+    ahead() {
+        assert(this._status);
+        return this._status === "ahead";
+    }
+
+    // feature <= base:
+    // either both branches point to same commit (e.g., just merged) or
+    // there are new commits on base since the feature branch was merged
+    merged() {
+        assert(this._status);
+        return this._status === "identical" || this._status === "behind";
+    }
+
+    // neither ahead nor merged:
+    // both base and the feature branch have unique-to-them commits
+    diverged() {
+        assert(this._status);
+        // this._status should be "diverged", but we also handle any new/unexpected statuses here
+        return !(this.ahead() || this.merged());
+    }
+}
+
+// a single GitHub pull request
+class PullRequest {
+
+    constructor(pr, banStaging) {
+        this._rawPr = pr; // may be rather old and lack pr.mergeable; see _loadRawPr()
+
+        this._shaLimit = 6; // how many SHA chars to show in debug messages
+
+        this._approval = null; // future Approval object
+
+        // optimization: cached _getRequiredContexts() result
         this._requiredContextsCache = null;
+
         this._tagSha = null;
-        this._role = null;
+        this._stagedPosition = null;
+        this._freshStagedCommitWithFailedChecks = null;
+
+        // optimization: cached _tagCommit() result
+        this._tagCommitCache = null;
+
+        // major methods we have called, in the call order (for debugging only)
+        this._breadcrumbs = [];
+
+        this._messageValid = null;
+
+        this._prState = null; // calculated PrState
+
+        this._labels = null;
+        // while unexpected, PR merging and closing is not prohibited when staging is
```

Comment:
To clarify the scope of this comment, please add new lines:
```suggestion

        // while unexpected, PR merging and closing is not prohibited when staging is
        this._stagingBanned = banStaging;
        
```